### PR TITLE
Release camelot-fork 0.20.1

### PR DESCRIPTION
**Pull request by foarsitter**
Leaving this here so the it can be merged if there is any interest.

### CODE REVIEW

**Comment by foarsitter** [comment]
This pull-requests does a lot of things. Since things are getting back to life over here I will explain them a bit more. 

1) The project is merged into  [cookiecutter-hypermodern-python](https://cookiecutter-hypermodern-python.readthedocs.io/en/2022.6.3.post1/) which comes with a lot of [features](https://cookiecutter-hypermodern-python.readthedocs.io/en/2022.6.3.post1/#features) and documented workflows for i.e. [how to release](https://cookiecutter-hypermodern-python.readthedocs.io/en/2022.6.3.post1/guide.html#how-to-make-a-release)

2) It removes [pdftopng](https://github.com/vinayak-mehta/pdftopng) as hard dependency since it is not compatible with python 3.9 and above. Tests that depend on poppler/pdftopng are skipped as long as pdftopng is not updated. The fork is compatible with 3.8 and above. It is up to discussion if we need to add more python versions. 

3) It aligns the type of PDFHandler.filepath with PdfReader.stream so Path and IO objects are accepted. `filepath` can now be an in memory file upload for example. 

4) I ran pre-commit to the whole branch to apply black & isort to the codebase. 

5) The first pull-requests with bugfixes are merged: https://github.com/foarsitter/camelot/releases/tag/v0.20.1 

From my point of view it makes sense to release version 1.0 as a stable release where we support ghostscript and do minor bugfixes and keep our dependencies up to date.

In version 2.0 we can add support for poppler/pdftopng.

What do you tink? Which changes can be cherry-picked and which changes need to dropped?

**Comment by ExSidius** [comment]
@foarsitter @vinayak-mehta do you have a timeline around when this will be merged in? 

I'm running into issues installing camelot on python 3.10 because of `pdftopng`.

Let me know if I can help at all.

**Comment by MartinThoma** [comment]
@foarsitter I would recommend to break a couple of changes out of this PR. That would have advantages:

1. It's more likely we get progress sooner
2. It's easier to review changes
3. Less merge conflicts

Especially black+isort would be VERY desirable to have in its own PR. That PR should not add / change anything else + it should document which commands were executed.

**Comment by MartinThoma** [comment]
@foarsitter https://github.com/camelot-dev/camelot/pull/358 applies only black

File: README.md
Comment (MartinThoma) [inline README.md:1]: that should definitely not go into the `master` branch

**Comment by vinayak-mehta** [comment]
This is a big one, let me go through it this weekend

File: README.md
Comment (foarsitter) [reply README.md]: True that, reverted changes to README.md. If this is getting merged (and that is a big if) we need to update the readme accordingly.

File: pyproject.toml
Comment (bosd) [inline pyproject.toml:1]: Please drop this from the pr, or update it to remove references to the fork.

**Review by bosd — CHANGES_REQUESTED** [review]
In general looks good to me. Few minor comments.

Only concern that this would possibly be incompatible with the multithreading.
But maybe thats a concern after we merged this.

**Comment by bosd** [comment]
Fixes #292

File: pyproject.toml
Comment (foarsitter) [reply pyproject.toml:1]: Thanks for the feedback, sloppy from my side.

File: .github/workflows/constraints.txt
Comment (foarsitter) [reply .github/workflows/constraints.txt:1]: Don't know exactly but the [tests](https://github.com/foarsitter/camelot/actions/runs/5561884461/jobs/10159842909) are succeeding so leaving it there for now.

**Review by bosd — APPROVED** [review]
LGTM now!

**Comment by bosd** [comment]
But the way, do you have the powers to publish a new release on pypi?

**Comment by foarsitter** [comment]
Thanks for your support @bosd, I think we finally can get things moving. 

We need one more maintainer te approve before we can merge this. Then we need to find out how we can release to pypi since I do not have the rights to edit the repository settings.

**Comment by bosd** [comment]
@vinayak-mehta Could you give @foarsitter access on Pypi? 
So he can publish the releases there.
Potentially setup the automatic pypi publisher from the cookiecutter project.
https://github.com/cjolowicz/cookiecutter-hypermodern-python

@maintainers Please review..

**Review by bosd — APPROVED** [review]

**Comment by MartinThoma** [comment]
> @vinayak-mehta Could you give @foarsitter access on Pypi?

Alternative, you could add secrets to this repository (which only you would be able to see) + add a github action. For example, when a git tag is added then the action is executed which triggers a build + release.

**Comment by foarsitter** [comment]
@MartinThoma a commit to master with a change in version in `pyproject.toml` will trigger the deploy. The configuration is [simple](https://cookiecutter-hypermodern-python.readthedocs.io/en/2022.6.3.post1/quickstart.html#pypi). 

> Sign up at [PyPI](https://pypi.org/).
> Go to the Account Settings on PyPI, generate an API token, and copy it.
> Go to the repository settings on GitHub, and add a secret named PYPI_TOKEN with the token you just copied.
> 
> TestPyPI
> Sign up at [TestPyPI](https://test.pypi.org/).
> Go to the Account Settings on TestPyPI, generate an API token, and copy it.
> Go to the repository settings on GitHub, and add a secret named TEST_PYPI_TOKEN with the token you just copied.

**Comment by bosd** [comment]
@MartinThoma Are you a maintainer here? Can you give the last approval to get this one in.
(This fixes gh actions, which is currently broken and blocking other PR's)

File: .github/workflows/release.yml
Comment (MartinThoma) [inline .github/workflows/release.yml:63]: Is this set?

File: LICENSE
Comment (MartinThoma) [inline LICENSE:3]: This is not OK. I will not accept the PR like this.

**Review by MartinThoma — CHANGES_REQUESTED** [review]
I want the license / copyright change to be explained / discussed.

**Comment by foarsitter** [comment]
Didn't know the cookiecutter template added a copyright with my name. Will remove it tomorrow. Thanks for checking.

**Comment by foarsitter** [comment]
Found some references at other places that needed to be updated. Sorry for the inconvenience.

File: LICENSE
Comment (foarsitter) [reply LICENSE]: Reverted

File: .github/workflows/release.yml
Comment (foarsitter) [reply .github/workflows/release.yml:63]: I do not have sufficient permissions to edit the secrets so I don't known

**Review by MartinThoma — APPROVED** [review]
Thank you for the many improvements :pray: 

I am not super familiar with camelot (and especially its CI). As a lot of files were touched (+4583 -975! and 145 files!) I recommend making some smoke-tests locally to ensure basic functionality.

In general, this would have been easier to review if the automatic changes (ruff? black? pyupgrade? isort?) were done in separate PRs.

**Comment by foarsitter** [comment]
The changes your are describing are in separated commits, you can select commits on the top left in the `Files changed` tab
image

What kind of smoke-tests do you have in mind? We are using the fork in production and all the tests of this PR are succeeding. Camelot has a pretty decent test coverage (88%) and a lot of different pdf types are tested: https://github.com/camelot-dev/camelot/tree/master/tests/files  

One of the reasons I'm using the hypermodern cookiecutter template is that it has a lot of documentation. In example you can read about all the Github actions this PR includes over [here](https://cookiecutter-hypermodern-python.readthedocs.io/en/2022.6.3.post1/guide.html#github-actions-workflows) and how to make a release over [here](https://cookiecutter-hypermodern-python.readthedocs.io/en/2022.6.3.post1/guide.html#how-to-make-a-release). 

I agree with you that the road I took is quite the spartan way but I think it is the fastest way forward.

**Comment by MartinThoma** [comment]
> The changes your are describing are in separated commits

Interesting. Most of the time I don't look at single commits as they contain a lot of noise. As a maintainer of pypdf, I also don't like this model as it "packages" different unrelated changes. Hence it removes the freedom of choosing which of those I want to take in and which I don't want. The result is typically that it takes overall longer to merge those PRs + causes more work.

In the case of camelot I can understand it though. It's been a while since the last release. Please take my comment here more as a general wish. It's not meant as criticism in this specific case.

> We are using the fork in production

Oh, nice! That should do it :-)

**Comment by bosd** [comment]
So.. concluding.. all lights are green :heart: It's a go :smiley: can we get a merge?

**Comment by MartinThoma** [comment]
Interesting, I thought you could merge @bosd . 
@foarsitter Can you merge? Do you know who can?

Just to clarify: I can merge, but not release to PyPI. I'm a bit hesitant as I would rather like to support once in a while than become a core maintainer of another big project :sweat_smile:

**Comment by foarsitter** [comment]
@MartinThoma yes I'm enable to merge. Since it is a large PR and a lot happened during last weekend I think it is appropriate to have a cooldown of a few days before merging so everyone can react. 

Releasing to `camelot` on pypi is something we need to solve. In the meantime I can release to `camelot-fork` by forking this repo.

**Comment by MartinThoma** [comment]
Sounds good!

Regarding the releases, I've opened https://github.com/camelot-dev/camelot/issues/389

**Comment by bosd** [comment]
> Interesting, I thought you could merge @bosd .
> Do you know who can?

No, I don't have any permissions here to merge.
I'm not listed as a maintainer (However I offered to be one).

As to my knowledge only @MartinThoma and @foarsitter are the only currently active maintainers.

**Comment by foarsitter** [comment]
Squash and merge is the only option for merging.... That need to be changed before merging.

**Comment by MartinThoma** [comment]
For this specific PR I agree. In general, I prefer squash-and-merge. PRs should only contain one change (once we are back on track)

**Comment by bosd** [comment]
> Squash and merge is the only option for merging.... That need to be changed before merging.

So this means that dev is blocked/stalled again?? Or does any of you have the super powers to change this?

**Comment by MartinThoma** [comment]
image

I don't have access to the settings.

**Comment by foarsitter** [comment]
As long there is nobody around with full permission for GitHub and pypi this project could be considered dead.

**Comment by MartinThoma** [comment]
@vinayak-mehta Can you please give me full access so that I can adjust the project settings (temporarily enabling a merge-commit for this specific PR)?

**Comment by MartinThoma** [comment]
@foarsitter It seems as if I will not get the necessary permissions any time soon. Would it be ok for you if I squash-merge this PR and then make a release to PyPI?

**Comment by foarsitter** [comment]
Sure, no objection if its the only way forward. Go ahead :)

**Comment by MartinThoma** [comment]
https://github.com/camelot-dev/camelot/blob/master/.github/workflows/release.yml seems to have an issue:

image

**Comment by MartinThoma** [comment]
"Poetry could not find a pyproject.toml file in /home/runner/work/camelot/camelot or its parents"

I don't know why this part is failing.

However, the PyPI part should be removed as it's very likely not configured.

**Comment by vinayak-mehta** [comment]
Sorry for the dropping the ball big time here, I thought you both had all the permissions because you were members of the github organization, I didn't know github doesn't let you change minor repo level permissions even if you're an org member

**Pull request merged**
